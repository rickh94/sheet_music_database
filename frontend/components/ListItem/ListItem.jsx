import React, { useState } from 'react'
import PropTypes from 'prop-types'

import Level from 'react-bulma-components/lib/components/level'
import Modal from 'react-bulma-components/lib/components/modal'
import Box from 'react-bulma-components/lib/components/box'
import Heading from 'react-bulma-components/lib/components/heading/heading'
import Button from 'react-bulma-components/lib/components/button'

import FieldDisplay from '../FieldDisplay'

function ListItem(props) {
  const [modalIsOpen, openModal] = useState(false)
  return (
    <React.Fragment>
      <Level>
        <Level.Side align="left">
          <Level.Item>
            <FieldDisplay
              value={props.item.name}
              saveCallback={props.saveCallback}
              backendFieldName=""
              errors={props.errors}
              linkTo={`/${props.link}/${props.item.id}`}
            />
          </Level.Item>
        </Level.Side>
        {props.deleteCallback && (
          <Level.Side align="right" className="absolutely-no-margin">
            <Level.Item>
              <div className="links">
                <a
                  onClick={e => {
                    e.preventDefault()
                    openModal(true)
                    // props.deleteCallback(props.item.id)
                  }}
                  className="edit-link"
                >
                  delete
                </a>
              </div>
            </Level.Item>
          </Level.Side>
        )}
      </Level>
      <Modal show={modalIsOpen} onClose={() => openModal(false)}>
        <Modal.Content>
          <Box>
            <div style={{ margin: '1rem' }}>
              <Heading size={3}>Confirm Delete</Heading>
              <p>Are you sure you want to delete {props.item.name}?</p>
              <Button
                color="danger"
                style={{ marginRight: '0.5rem' }}
                onClick={e => {
                  e.preventDefault()
                  props.deleteCallback(props.item.id)
                  openModal(false)
                }}
              >
                Confirm
              </Button>
              <Button
                onClick={e => {
                  e.preventDefault()
                  openModal(false)
                }}
              >
                Cancel
              </Button>
            </div>
          </Box>
        </Modal.Content>
      </Modal>
    </React.Fragment>
  )
}

ListItem.propTypes = {
  item: PropTypes.object.isRequired,
  saveCallback: PropTypes.func,
  deleteCallback: PropTypes.func,
  errors: PropTypes.string,
  link: PropTypes.string
}

export default ListItem
