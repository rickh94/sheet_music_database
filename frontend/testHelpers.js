import React from 'react'

export const testToken = 'testtoken'

export function testLoggedIn(wrapper) {
  const history = { push: jest.fn() }
  const token = null
  const alert = { show: jest.fn() }
  wrapper.setProps({ token, history, alert })
  wrapper.instance().componentDidMount()
  expect(history.push).toHaveBeenCalledWith('/login')
  expect(alert.show).toHaveBeenCalledWith(
    <span className="alert-text">Please log in</span>,
    { type: 'error' }
  )
}

export function testGetData(wrapper, functionPropName, functionPropArgs = []) {
  const getFunction = jest.fn()
  wrapper.setProps({ [functionPropName]: getFunction, token: testToken })
  wrapper.instance().componentDidMount()
  expect(getFunction).toHaveBeenCalledWith(testToken, ...functionPropArgs)
}

export function testGetDataFailed(wrapper, functionPropName, nameInMessage) {
  const getFunction = () => false
  const alert = { show: jest.fn() }
  wrapper.setProps({ [functionPropName]: getFunction, token: testToken, alert })
  wrapper.instance().componentDidMount()
  expect(alert.show).toHaveBeenCalledWith(
    <span className="alert-text">{`Could not get ${nameInMessage}`}</span>,
    { type: 'error' }
  )
}

export function clickThing(tag, wrapper, content) {
  wrapper
    .findWhere(el => el.name() == tag && el.contains(content))
    .simulate('click', { preventDefault: jest.fn() })
}

export const clickButton = (wrapper, content) => clickThing('Button', wrapper, content)
export const clickLink = (wrapper, content) => clickThing('a', wrapper, content)

