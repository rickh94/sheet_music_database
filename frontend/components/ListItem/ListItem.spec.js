import React from 'react'
import { shallow } from 'enzyme'
import ListItem from './ListItem'

import '../../setupTests'
import { clickButton } from '../../testHelpers';

describe('TagItem', () => {
  const testItem = { id: 5, name: 'test' }
  let wrapper
  beforeEach(() => {
    wrapper = shallow(
      <ListItem
        item={testItem}
        saveCallback={jest.fn()}
        deleteCallback={jest.fn()}
        link={'test item'}
      />
    )
  })

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a Level', () => {
    expect(wrapper.exists('Level')).toBeTruthy()
  })

  it('has delete link with icon', () => {
    expect(wrapper.contains('delete')).toBeTruthy()
  })

  it('does not render a delete link with no delete callback', () => {
    wrapper.setProps({deleteCallback: null})
    expect(wrapper.contains('delete')).toBeFalsy()
  })

  it('has a confirm delete modal', () => {
    expect(wrapper.find('Modal').contains('Confirm Delete')).toBeTruthy()
  })

  it('calls passed in function when confirm is clicked', () => {
    const deleteCallback = jest.fn()
    wrapper.setProps({ deleteCallback })
    clickButton(wrapper, 'Confirm')
    expect(deleteCallback).toHaveBeenCalledWith(testItem.id)
  })
})
