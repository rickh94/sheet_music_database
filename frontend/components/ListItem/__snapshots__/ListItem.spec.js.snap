// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagItem matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ListItem
    deleteCallback={[MockFunction]}
    item={
      Object {
        "id": 5,
        "name": "test",
      }
    }
    link="test item"
    saveCallback={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "backgroundColor": undefined,
      "breakpoint": null,
      "children": Array [
        <FieldDisplay
          backendFieldName=""
          linkTo="/test item/5"
          saveCallback={[MockFunction]}
          value="test"
        />,
        <div
          className="links"
        >
          <a
            className="edit-link"
            onClick={[Function]}
          >
            delete
          </a>
        </div>,
      ],
      "className": "",
      "clearfix": undefined,
      "clipped": undefined,
      "hidden": undefined,
      "invisible": undefined,
      "italic": undefined,
      "marginless": undefined,
      "overlay": undefined,
      "paddingless": undefined,
      "pull": undefined,
      "radiusless": undefined,
      "renderAs": "div",
      "responsive": undefined,
      "shadowless": undefined,
      "style": Object {},
      "textAlignment": undefined,
      "textColor": undefined,
      "textSize": undefined,
      "textTransform": undefined,
      "textWeight": undefined,
      "unselectable": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "backendFieldName": "",
          "errors": undefined,
          "linkTo": "/test item/5",
          "saveCallback": [MockFunction],
          "value": "test",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <a
            className="edit-link"
            onClick={[Function]}
          >
            delete
          </a>,
          "className": "links",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "delete",
            "className": "edit-link",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "delete",
          "type": "a",
        },
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "backgroundColor": undefined,
        "breakpoint": null,
        "children": Array [
          <FieldDisplay
            backendFieldName=""
            linkTo="/test item/5"
            saveCallback={[MockFunction]}
            value="test"
          />,
          <div
            className="links"
          >
            <a
              className="edit-link"
              onClick={[Function]}
            >
              delete
            </a>
          </div>,
        ],
        "className": "",
        "clearfix": undefined,
        "clipped": undefined,
        "hidden": undefined,
        "invisible": undefined,
        "italic": undefined,
        "marginless": undefined,
        "overlay": undefined,
        "paddingless": undefined,
        "pull": undefined,
        "radiusless": undefined,
        "renderAs": "div",
        "responsive": undefined,
        "shadowless": undefined,
        "style": Object {},
        "textAlignment": undefined,
        "textColor": undefined,
        "textSize": undefined,
        "textTransform": undefined,
        "textWeight": undefined,
        "unselectable": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "backendFieldName": "",
            "errors": undefined,
            "linkTo": "/test item/5",
            "saveCallback": [MockFunction],
            "value": "test",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <a
              className="edit-link"
              onClick={[Function]}
            >
              delete
            </a>,
            "className": "links",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "delete",
              "className": "edit-link",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "delete",
            "type": "a",
          },
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
